src/Provider/AbstractProvider.php

```diff
@@ -674,6 +674,10 @@ protected function parseResponse(ResponseInterface $response)
                 throw $e;
             }
 
+            if ($response->getStatusCode() == 500) {
+                throw new UnexpectedValueException('Failed to parse JSON and OAuth server response code is 500. It could be internal error on Oauth server.');
```

Comment:
@ramsey I'm not opposed to this change. It feels a little worrisome to me that we would single out a specific error code, but in this situation it probably makes sense.

The only thing I would like to see changed is passing the current `$e` as `$previousException` to the newly added exception.